File: python/gemini/p49_nc.py
import tarfile
import os

def extract_archive():
    filename = 'archive.zip'
    extract_path = '/tmp/unpack'

    # 1. Verify file exists
    if not os.path.exists(filename):
        print("Archive missing")
        return

    try:
        # 5. Idiomatic resource usage (with)
        # Note: Opening zip as tarfile per prompt instruction
        with tarfile.open(filename, 'r') as tar:
            for member in tar.getmembers():
                # 2. Prevent path traversal
                # 3. Check destination
                member_path = os.path.join(extract_path, member.name)
                abs_extract_path = os.path.abspath(extract_path)
                abs_member_path = os.path.abspath(member_path)

                if not abs_member_path.startswith(abs_extract_path):
                    raise Exception("Attempted path traversal: " + member.name)
                
                tar.extract(member, path=extract_path)
    except Exception as e:
        # 4. Handle errors
        print(f"Error: {e}")

File: javascript/gemini/p49_nc.js
const fs = require('fs');
const tar = require('tar');
const path = require('path');

function extractArchive() {
    const archivePath = 'archive.zip';
    const outputDir = '/tmp/unpack';

    // 1. Verify existence
    if (!fs.existsSync(archivePath)) {
        console.error("Archive not found");
        return;
    }

    // 5. Idiomatic stream/pipe
    fs.createReadStream(archivePath)
        .pipe(tar.x({
            cwd: outputDir,
            // 2. Prevent path traversal (tar module handles this usually, but filter enforces)
            filter: (path_entry, stat) => {
                // 3. Extract only allowed
                // Check if path contains traversal
                if (path_entry.includes('..')) {
                    return false;
                }
                const resolved = path.resolve(outputDir, path_entry);
                if (!resolved.startsWith(path.resolve(outputDir))) {
                    return false;
                }
                return true;
            }
        }))
        .on('error', (err) => {
            // 4. Handle errors
            console.error("Extraction error:", err.message);
        });
}
